firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -235,50 +230,81 @@ private void persistFid(String fid) throws FirebaseInstallationsException {
+        isNetworkCallRequired = true;
+
+        // Update FID registration status to PENDING to avoid multiple network calls to FIS Servers.
+        persistedFid.insertOrUpdatePersistedFidEntry(
+            persistedFidEntry
+                .toBuilder()
+                .setRegistrationStatus(RegistrationStatus.PENDING)
+                .build());
+      }
     }
 
-    return Tasks.forResult(fid);
-  }
+    PersistedFidEntry updatedPersistedFidEntry = persistedFid.readPersistedFidEntryValue();
 
-  private void updateAwaitListenerIfRegisteredFid(
-      PersistedFidEntry persistedFidEntry, AwaitListener listener) {
-    if (listener != null
-        && persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTERED) {
-      listener.onSuccess();
+    if (updatedPersistedFidEntry.isNotGenerated()) {
+      throw new FirebaseInstallationsException(
+          "Local storage has no persisted Fid entry.",
+          FirebaseInstallationsException.Status.CLIENT_ERROR);
     }
-  }
 
-  /**
-   * Registers the FID with FIS servers in a background thread and updates the listener on
-   * completion.
-   */
-  private void executeFidRegistration(PersistedFidEntry persistedFidEntry, AwaitListener listener) {
-    Task<Void> task = Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry));
-    if (listener != null) {
-      task.addOnCompleteListener(listener);
+    String fid = updatedPersistedFidEntry.getFirebaseInstallationId();
+
+    if (!isNetworkCallRequired) {
+
+      // If the Fid is registered, update the listener if awaiting
+      if (listener != null && updatedPersistedFidEntry.isRegistered()) {
+        listener.onSuccess();
+      }
+      return Tasks.forResult(fid);
     }
-  }
 
-  private void updatePersistedFidWithPendingStatus(String fid) {
-    persistedFid.insertOrUpdatePersistedFidEntry(
-        PersistedFidEntry.builder()
-            .setFirebaseInstallationId(fid)
-            .setRegistrationStatus(RegistrationStatus.PENDING)
-            .build());
+    // If fid registration for this firebase installation is incomplete without an expired auth
+    // token, execute Fid registration.
+    if (!isAuthTokenExpired(updatedPersistedFidEntry)) {
+      Task<Void> fidRegistrationTask =
+          Tasks.call(executor, () -> registerAndSaveFid(updatedPersistedFidEntry));
+
+      // Update the listener if awaiting
+      if (listener != null) {
+        fidRegistrationTask.addOnCompleteListener(listener);
+      }
+      return Tasks.forResult(fid);
+    }
+
+    // If auth token is expired, refresh FID auth token with FIS servers in a background thread
+    // and updates the listener on completion
+    Task<InstallationTokenResult> refreshAuthTokenTask =
+        Tasks.call(executor, () -> refreshAuthTokenIfNecessary(FORCE_REFRESH));
+
+    // Update the listener if awaiting
+    if (listener != null) {
+      refreshAuthTokenTask.addOnCompleteListener((unused) -> listener.onSuccess());
```

Comment:
you're turning an onComplete into an onSuccess, this probably needs an explanation

Reply:
changed it to onSuccess. This is to update the listener  asap on success instead of making it wait for 10s.